src/providers/postgres/qgspostgresexpressioncompiler.cpp

```diff
@@ -32,3 +36,91 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
   return QgsPostgresConn::quotedValue( value );
 }
 
+static const QMap<QString, QString>& functionNamesSqlFunctionsMap()
+{
+  static QMap<QString, QString> fnNames;
+  if ( fnNames.isEmpty() )
+  {
+    // functions
+    fnNames.insert( QStringLiteral( "disjoint" ), QStringLiteral( "ST_Disjoint" ) );
+    fnNames.insert( QStringLiteral( "intersects" ), QStringLiteral( "ST_Intersects" ) );
+    fnNames.insert( QStringLiteral( "touches" ), QStringLiteral( "ST_Touches" ) );
+    fnNames.insert( QStringLiteral( "crosses" ), QStringLiteral( "ST_Crosses" ) );
+    fnNames.insert( QStringLiteral( "contains" ), QStringLiteral( "ST_Contains" ) );
+    fnNames.insert( QStringLiteral( "overlaps" ), QStringLiteral( "ST_Overlaps" ) );
+    fnNames.insert( QStringLiteral( "within" ), QStringLiteral( "ST_Within" ) );
+    fnNames.insert( QStringLiteral( "translate" ), QStringLiteral( "ST_Translate" ) );
+    fnNames.insert( QStringLiteral( "buffer" ), QStringLiteral( "ST_Buffer" ) );
+  }
+  return fnNames;
+}
+
+static QString sqlFunctionFromFunctionName( const QString& fnName )
+{
+  return functionNamesSqlFunctionsMap().value( fnName, QString() );
+}
+
+QgsSqlExpressionCompiler::Result QgsPostgresExpressionCompiler::compileNode( const QgsExpression::Node* node, QString& result )
+{
+  switch ( node->nodeType() )
+  {
+    case QgsExpression::ntFunction:
+    {
+      const QgsExpression::NodeFunction* n = static_cast<const QgsExpression::NodeFunction*>( node );
+
+      QgsExpression::Function* fd = QgsExpression::Functions()[n->fnIndex()];
+      if ( fd->name() == "$geometry" )
+      {
+        result = quotedIdentifier( mGeometryColumn );
+        return Complete;
+      }
+
+      QStringList args;
+
+      Result inResult = Complete;
+      Q_FOREACH ( const QgsExpression::Node* ln, n->args()->list() )
+      {
+        QString s;
+        Result r = compileNode( ln, s );
+        if ( r == Complete || r == Partial )
+        {
+          args << s;
+          if ( r == Partial )
+            inResult = Partial;
+        }
+        else
+          return r;
+      }
+
+      QString nd;
+      if ( fd->name() == "geom_from_wkt" )
+      {
+        nd = QStringLiteral( "ST_GeomFromText" );
+        args << ( mRequestedSrid.isEmpty() ? mDetectedSrid : mRequestedSrid );
+      }
+      else if ( fd->name() == "geom_from_gml" )
+      {
+        nd = QStringLiteral( "ST_GeomFromGML" );
+        args << ( mRequestedSrid.isEmpty() ? mDetectedSrid : mRequestedSrid );
+      }
+      else
+      {
+        nd = sqlFunctionFromFunctionName( fd->name() );
```

Comment:
It'd be a good idea to move sqlFunctionFromFunctionName and use of this up to the base QgsSqlExpressionCompiler class. This would make it easy to reuse for other providers.